Web/Scotty/Action.hs

```diff
@@ -354,6 +363,19 @@ jsonData = do
         A.Error err -> throwIO $ FailedToParseJSON b $ T.pack err
         A.Success a -> return a
 
+-- | Parse the request body as @x-www-form-urlencoded@ form data and return it.
+--
+--   The form is parsed using 'urlDecodeAsForm'. If that returns 'Left', the
+--   status is set to 400 and an exception is thrown.
+--
+-- NB : Internally this uses 'body'.
+formData :: (FromForm a, MonadIO m) => ActionT m a
+formData = do
+  b <- body
```

Comment:
Does `body` preserve the request body across calls to `next`? see #308

Reply:
I'm not 100% what the desired behavior is, I assumed this was "right" because of the way `jsonData` is written :)

Should I add a test here?

Reply:
![image](https://github.com/scotty-web/scotty/assets/5812813/21a274e5-ec53-4512-be18-05c817fa8c16)

Internally, both `formParams` and `formData` use `getBodyAction` which I believe leads to consistent behavior unless I'm missing something.

Reply:
Correct! (beautiful diagram btw, how did you make it?) 

Basically this "lazier" parsing of request bodies (underneath the implementation of `files`, `formParams` etc.) is what I introduced in #369 and I think it would make sense for form parameters too.

Reply:
The diagram is made using [plant UML](https://plantuml.com), basically a text-based UML tool. Getting the call graph from Haskell isn't automatic tho, I did that by hand. Below is the source code for it.

<details>

```plantuml
@startuml
skinparam dpi 300

hide members
hide circle

package "Web.Scotty.Action" {
    formParams -down-> "Web.Scotty.Internal.Types.formParamsAndFilesWith"
    formData -down-> body
    body -down-> "Web.Scotty.Internal.Types.envBody"
}

package "Web.Scotty.Internal.Types" {
    formParamsAndFilesWith -left-> envFormDataAction
    envFormDataAction -down-> "Web.Scotty.Body.getFormParamsAndFilesAction"
    envBody -down-> Web.Scotty.Body.getBodyAction
}

note top of Web.Scotty.Internal.Types.envFormDataAction
Initialized by mkEnv
end note

note top of Web.Scotty.Internal.Types.envBody
Initialized by mkEnv
end note

package "Web.Scotty.Body" {
    getFormParamsAndFilesAction -down-> getBodyAction
    getFormParamsAndFilesAction -right-> parseRequestBodyExBS
    parseRequestBodyExBS -down-> "Network.Wai.Parse.sinkRequestBodyEx"
}

package "Network.Wai.Parse" {
    sinkRequestBodyEx -down-> conduitRequestBodyEx
    conduitRequestBodyEx -down-> "Network.HTTP.Types.parseSimpleQuery"
}

package Network.HTTP.Types {
    parseSimpleQuery -right-> parseQuery
    parseQuery -right-> urlDecode
}
@enduml
```

</details>